src/ucs/datastruct/usage_tracker.c

```diff
@@ -0,0 +1,244 @@
+}
+
+/* Checks if an entry has high enough score to get promoted. */
+static int
+ucs_usage_tracker_compare(const void *elem1, const void *elem2, void *arg)
+{
+    const ucs_usage_tracker_params_t *params = arg;
+    ucs_usage_tracker_element_t *item1 = *(ucs_usage_tracker_element_t**)elem1;
+    ucs_usage_tracker_element_t *item2 = *(ucs_usage_tracker_element_t**)elem2;
+
+    if (fabs(item1->score - item2->score) < params->remove_thresh) {
+        return (item1->key > item2->key) ? 1 : -1;
+    }
+
+    return (item1->score < item2->score) ? 1 : -1;
+}
+
+/* Promote/Demote entires base on the latest score, and triggers user
+  * callback accordingly. */
+static void ucs_usage_tracker_update_rank(ucs_usage_tracker_h usage_tracker)
+{
+    const ucs_usage_tracker_params_t *params = &usage_tracker->params;
+    size_t elem_index                        = 0;
+    ucs_usage_tracker_element_t **elems_array, *item;
+    khiter_t iter;
+    uint64_t key;
+
+    elems_array = ucs_alloca(sizeof(*elems_array) *
+                             (kh_size(&usage_tracker->hash)));
+
+    kh_foreach_key(&usage_tracker->hash, key,
+        iter                    = kh_get(usage_tracker_hash,
+                                         &usage_tracker->hash, key);
+        item                    = &kh_val(&usage_tracker->hash, iter);
+        elems_array[elem_index] = item;
+        elem_index ++;
+    )
+
+        qsort_r(elems_array, kh_size(&usage_tracker->hash),
+                sizeof(*elems_array), ucs_usage_tracker_compare,
+                (ucs_usage_tracker_params_t*)params);
+
+    for (elem_index = 0; elem_index < kh_size(&usage_tracker->hash);
+         ++elem_index) {
+        item = elems_array[elem_index];
+
+        if (elem_index < params->promote_thresh) {
+            if (item->promoted) {
+                continue;
+            }
+
+            params->promote_cb(item->key, params->promote_arg);
+            item->promoted = 1;
+        }
+
+        else if (elem_index >= params->promote_capacity) {
+            if (item->promoted) {
+                params->demote_cb(item->key, params->demote_arg);
+                item->promoted = 0;
+            }
+
+            ucs_usage_tracker_remove(usage_tracker, item->key);
+        }
+    }
+}
+
+void ucs_usage_tracker_progress(ucs_usage_tracker_h usage_tracker)
+{
+    void **item;
+    khiter_t iter;
+    ucs_usage_tracker_element_t *elem;
+    uint64_t key;
+
+    kh_foreach_key(&usage_tracker->hash, key,
+        iter = kh_get(usage_tracker_hash, &usage_tracker->hash, key);
+        elem = &kh_val(&usage_tracker->hash, iter);
+
+        elem->score *= usage_tracker->params.exp_decay.a;
+        if (ucs_lru_has_entry(usage_tracker->lru, (void *)key)) {
+            elem->score += usage_tracker->params.exp_decay.b;
+        }
+    )
+
+    ucs_lru_for_each(item, usage_tracker->lru) {
+        ucs_usage_tracker_put(usage_tracker, *item);
+    }
+
+    if (kh_size(&usage_tracker->hash) > 0) {

```

Comment:
why need to check the size?

Reply:
1. rename ucs_usage_tracker_update_rank to ucs_usage_tracker_promote
2. the function itself can exit if elems_count==0